README
NAME
    Catalyst::TraitFor::Model::DBIC::Shortcut - shortcuts support for DBIC
    models

VERSION
    Version 0.02

SYNOPSIS
        #
        # application class
        #
        package TestApp;

        use Moose;
        use namespace::autoclean;
        use Catalyst qw/ ......... /;
        extends 'Catalyst';
        with 'Catalyst::TraitFor::Model::DBIC::Shortcut';


        #
        # controller class
        #
        package TestApp::Controller::Test;

        .........
            # these two calls are the same
            my $s = $c->model('DB')->schema;
            my $s = $c->db_schema;
        .........
            # these two calls are the same
            my $rs = $c->model('DB::Actor');
            my $rs = $c->db_actor_rs;
        .........

DESCRIPTION
    If you got tired of writting "$c->model('DB::Actor')" each time, or if
    you use auto-completion intensively, you could look at this trait. Just
    use this role in your application class, and you'll have shortcuts
    auto-created for all DBIx::Class-based models:

    - schema
        for all schema classes, based on Catalyst::Model::DBIC::Schema,
        you'll get method with name "lowercase class name" + "_schema", with
        all "::" converted to underscore ("_"):

            $c->model('DB')->schema         ==>     $c->db_schema
            $c->model('DBIC')->schema       ==>     $c->dbic_schema
            $c->model('DBIC::DB1')->schema  ==>     $c->dbic_db1_schema

    - resultset
        for all resultset classes, based on DBIx::Class, you'll get method
        with name "lowercase class name" + "_rs", with all "::" converted to
        underscore ("_"):

            $c->model('DB::Actor')          ==>     $c->db_actor_rs
            $c->model('DB::Track')          ==>     $c->db_track_rs
            $c->model('DBIC::DB1::Actor')   ==>     $c->dbic_db1_actor_rs

GIT REPOSITORY
    git clone
    git://github.com/cub-uanic/Catalyst-TraitFor-Model-DBIC-Shortcut.git

SEE ALSO
    Catalyst, DBIx::Class, Moose

AUTHOR
    Oleg Kostyuk, "<cub#cpan.org>"

    Based on ideas from from Pedro Melo and Oleg Pronin

    <http://lists.scsys.co.uk/pipermail/dbix-class/2010-January/008794.html>

    <http://lists.scsys.co.uk/pipermail/dbix-class/2010-February/008903.html
    >

COPYRIGHT & LICENSE
    Copyright by Oleg Kostyuk.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Catalyst-TraitFor-Model-DBIC-Shortcut

0.02    2010-06-20 02:08:12
        Switch repository back to Github
        Small spelling corrections

0.01    2010-02-08 17:57:40
        First version, released on an unsuspecting world :)



Makefile.PL
use strict;
use warnings;
use inc::Module::Install;
use Module::Install::AuthorTests;
use Module::Install::AuthorRequires;

license 'perl';

name 'Catalyst-TraitFor-Model-DBIC-Shortcut';
all_from 'lib/Catalyst/TraitFor/Model/DBIC/Shortcut.pm';
author q{Oleg Kostyuk <cub@cpan.org>};

requires 'Catalyst'                      => '5.80';
requires 'Catalyst::Model::DBIC::Schema' => '0.31';
requires 'Moose';
requires 'Moose::Role';
requires 'namespace::autoclean';

test_requires 'Test::More';

author_requires 'Test::EOL' => '0.3';
author_requires 'Test::NoTabs';
author_requires 'Test::Pod'           => '1.14';
author_requires 'Test::Pod::Coverage' => '1.08';

resources repository => 'git://github.com/cub-uanic/Catalyst-TraitFor-Model-DBIC-Shortcut.git';

tests 't/*.t';
author_tests 't/author';

if ($Module::Install::AUTHOR) {
    system("pod2text lib/Catalyst/TraitFor/Model/DBIC/Shortcut.pm > README") and die $!;
}

WriteAll;



lib/Catalyst/TraitFor/Model/DBIC/Shortcut.pm
package Catalyst::TraitFor::Model::DBIC::Shortcut;

use namespace::autoclean;
use Moose::Role;

=head1 NAME

Catalyst::TraitFor::Model::DBIC::Shortcut - shortcuts support for DBIC models

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';

=head1 SYNOPSIS

    #
    # application class
    #
    package TestApp;

    use Moose;
    use namespace::autoclean;
    use Catalyst qw/ ......... /;
    extends 'Catalyst';
    with 'Catalyst::TraitFor::Model::DBIC::Shortcut';


    #
    # controller class
    #
    package TestApp::Controller::Test;

    .........
        # these two calls are the same
        my $s = $c->model('DB')->schema;
        my $s = $c->db_schema;
    .........
        # these two calls are the same
        my $rs = $c->model('DB::Actor');
        my $rs = $c->db_actor_rs;
    .........


=head1 DESCRIPTION

If you got tired of writting C<< $c->model('DB::Actor') >> each time, or if
you use auto-completion intensively, you could look at this trait.
Just use this role in your application class, and you'll have shortcuts
auto-created for all L<DBIx::Class>-based models:

=over 4

=item - schema

for all schema classes, based on L<Catalyst::Model::DBIC::Schema>, you'll get
method with name "lowercase class name" + "_schema", with all "::" converted to
underscore ("_"):

    $c->model('DB')->schema         ==>     $c->db_schema
    $c->model('DBIC')->schema       ==>     $c->dbic_schema
    $c->model('DBIC::DB1')->schema  ==>     $c->dbic_db1_schema

=item - resultset

for all resultset classes, based on L<DBIx::Class>, you'll get
method with name "lowercase class name" + "_rs", with all "::" converted to
underscore ("_"):

    $c->model('DB::Actor')          ==>     $c->db_actor_rs
    $c->model('DB::Track')          ==>     $c->db_track_rs
    $c->model('DBIC::DB1::Actor')   ==>     $c->dbic_db1_actor_rs

=back

=cut

after 'setup_finalize' => sub {
    my $self = shift;

    for my $m ( sort $self->models ) {
        my $model = $self->model($m);
        my $acc   = lc($m);
        $acc =~ s/::/_/g;

        if ( $model->isa('Catalyst::Model::DBIC::Schema') ) {
            $acc .= '_schema';

            if ( $self->can($acc) ) {
                ## log
                next;
            }
            $self->mk_group_accessors( inherited => $acc );
            $self->$acc( $model->schema );
        }
        elsif ( $model->isa('DBIx::Class') ) {
            $acc .= '_rs';
            if ( $self->can($acc) ) {
                ## log
                next;
            }
            $self->mk_group_accessors( inherited => $acc );
            $self->$acc($$model);
        }
    }

};

__PACKAGE__->meta->make_immutable();

=head1 GIT REPOSITORY

git clone git://github.com/cub-uanic/Catalyst-TraitFor-Model-DBIC-Shortcut.git

=head1 SEE ALSO

L<Catalyst>, L<DBIx::Class>, L<Moose>

=head1 AUTHOR

Oleg Kostyuk, C<< <cub#cpan.org> >>

Based on ideas from from Pedro Melo and Oleg Pronin

L<http://lists.scsys.co.uk/pipermail/dbix-class/2010-January/008794.html>

L<http://lists.scsys.co.uk/pipermail/dbix-class/2010-February/008903.html>


=head1 COPYRIGHT & LICENSE

Copyright by Oleg Kostyuk.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;    # End of Catalyst::TraitFor::Model::DBIC::Shortcut

